src/adapter/graphql/src/queries/datasets/dataset_endpoints.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use opendatafabric as odf;
+use url::Url;
+
+use crate::prelude::*;
+use crate::queries::*;
+
+pub struct DatasetEndpoints<'a> {
+    owner: &'a Account,
+    dataset_handle: odf::DatasetHandle,
+}
+
+#[Object]
+impl<'a> DatasetEndpoints<'a> {
+    #[graphql(skip)]
+    fn get_base_url() -> std::result::Result<Url, InternalError> {
+        // TODO: Use value from config not envvar
+        //       https://github.com/kamu-data/kamu-node/issues/45
+        let raw_base_url =
+            std::env::var("KAMU_BASE_URL").unwrap_or_else(|_| "http://127.0.0.1:8080".to_string());
+
+        Url::parse(&raw_base_url).int_err()
+    }
+
+    #[graphql(skip)]
+    pub fn new(owner: &'a Account, dataset_handle: odf::DatasetHandle) -> Self {
+        Self {
+            owner,
+            dataset_handle,
+        }
+    }
+
+    #[allow(clippy::unused_async)]
+    async fn web_link(&self) -> Result<LinkProtocolDesc> {
+        let base_url = Self::get_base_url()?;
+        let url = format!("{base_url}{}", self.dataset_handle.alias);
+
+        Ok(LinkProtocolDesc { url })
+    }
+
+    #[allow(clippy::unused_async)]
+    async fn cli(&self) -> Result<CliProtocolDesc> {
+        let base_url = Self::get_base_url()?;
+        let url = format!("{base_url}{}", self.dataset_handle.alias);
+
+        let push_command = format!("kamu push {url}");
+        let pull_command = format!(
+            "kamu pull {url} --as {}",
+            self.dataset_handle.alias.dataset_name
+        );
+
+        Ok(CliProtocolDesc {
+            pull_command,
+            push_command,
+        })
+    }
+
+    #[allow(clippy::unused_async)]
+    async fn rest(&self) -> Result<RestProtocolDesc> {
+        let base_url = Self::get_base_url()?;
+        let url = format!("{base_url}{}", self.dataset_handle.alias);
+
+        let tail_url = format!("{url}/tail?limit=10");
+        let query_url = format!("{base_url}graphql?query=query {{%0A%20 apiVersion%0A}}%0A");
+        let push_url = format!("{url}/push");
+
+        Ok(RestProtocolDesc {
+            tail_url,
+            query_url,
+            push_url,
+        })
+    }
+
+    #[allow(clippy::unused_async)]
+    async fn flightsql(&self) -> Result<FlightSqlDesc> {
+        // TODO: replace hardcoded value
+        Ok(FlightSqlDesc {
+            url: "datafusion+flightsql://node.demo.kamu.dev:50050".to_string(),
```

Comment:
I suggest a separate env per protocol that lives on a different port:
- `KAMU_BASE_URL_REST=https://node.kamu.dev`
- `KAMU_BASE_URL_FLIGHTSQL=grpc+tls://node.kamu.dev:50050`

Reply:
URL conventions for FlightSQL are in a messy state right now, but for now in our deployments let's use this style:
- `grpc://localhost:50050` - no TLS
- `grpc+tls://node.kamu.dev:50050` - with TLS

This is how [ADBC client](https://github.com/kamu-data/kamu-cli/blob/master/examples/flight-sql/python/client_flightsql_adbc.py) connects (which seems to be the most mature and widely used one).

Reply:
1) We should not mention `node.demo.kamu.dev` explicitly. This address should be externally configurable.
2) Regarding ports - I'm not 100% sure, but take a look how ports are defined when initializing TCP listeners in each correspoinding server. I think at least some a hardcoded now. If we assume external configuration, it should be done identically both here and in initialization code.

Reply:
Think of an utility helper, separate from this GraphQL query, that would centralize access to address and port information for each protocol. It could be used from here and all other code places that require URL and ports. If we have a particular configuration mechanism, like env vars, centralized in one place for all URLs and ports, it would be much easier to maintain when we apply different mechanism (YAML).

Reply:
> I suggest a separate env per protocol that lives on a different port:
> KAMU_BASE_URL_REST=https://node.kamu.dev
> KAMU_BASE_URL_FLIGHTSQL=grpc+tls://node.kamu.dev:50050

Done

---

> grpc://localhost:50050 - no TLS

Used as default value for `KAMU_BASE_URL_FLIGHTSQL`

> grpc+tls://node.kamu.dev:50050 - with TLS
> ^^^^^^^

TLS variant can be set via Config

---

> Think of an utility helper, separate from this GraphQL query

Done